First host implementation.

adbh94 opened the pull request:
Hosts consist of two internal processes: monitor_outgoing_packets and monitor_incoming_packets. These follow a passivate/reactivate framework through the events send_packet and receive_packet, respectively. Note that the host class itself performs the reactivations of its two internal processes. Flows and links indirectly reactivate the internal host processes through the send_packet and receive_packet API, respectively.

Is this a suitable framework for process interaction? I like this better than interruptions since one cannot pass objects through interruptions (I think). If this looks good, let's commit to this model for future process interactions.

Tests to follow upon implementation of environment, flows, and links.

adbh94 committed:
First host implementation.

Hosts consist of two internal processes: monitor_outgoing_packets and
monitor_incoming_packets. These follow a passivate/reactivate framework
through the events send_packet and receive_packet, respectively. Note
that the host class itself performs the reactivations of its two
internal processes. Flows and links indirectly reactivate the internal
host processes through the send_packet and receive_packet API,
respectively.

Is this a suitable framework for process interaction? I like this better
than interruptions since one cannot pass objects through interruptions
(I think). If this looks good, let's commit to this model for future
process interactions.

Tests to follow upon implementation of environment, flows, and links.

adbh94 committed:
Indentation nit.

adbh94 committed:
Corrected not operator

sdhpzhtk commented on host.py:
```diff
@@ -0,0 +1,170 @@
+        self.flows = flows
+        
+        # Set up packet buffers and notification events.
+        self.outgoing_packets = []
```
You have a buffer for outgoing_packets, so this is for alloha?

adbh94 replied:
We need a buffer to handle collisions. The host will know if there is a collision if it finds more than one packet in its outgoing_packets buffer.

sdhpzhtk commented on host.py:
```diff
@@ -0,0 +1,170 @@
+        
+        # Set up packet buffers and notification events.
+        self.outgoing_packets = []
+        self.incoming_packets = []
```
Why do you need a buffer for incoming_packet? There is only one link.... so it is impossible to have multiple packets simultaneously.

adbh94 replied:
You're right, but an incoming_packets buffer adds generality and mirrors the code for outgoing packets.

sdhpzhtk commented on host.py:
```diff
@@ -0,0 +1,170 @@
+        
+    def receive_packet(self, incoming_packet):
+        """Method called by link to transmit packet into the host."""
+
```
I do not think you need to use buffer for receiving packets.

sdhpzhtk commented:
You also need to collect stats in the host?! And write the report function that returns all the stats... though I feel like we should first discuss the format and type of data we are collecting....

adbh94 commented:
Yes, the statistics reporting function still needs to be implemented. I'm not sure right now exactly what we want to measure. Perhaps next meeting let's actually enumerate all the statistics we want to collect from each process.

kexinrong commented on host.py:
```diff
@@ -0,0 +1,170 @@
+        self.env = env
+        self.host_id = host_id
+        self.link = link
+        self.flows = flows
```
Quick question: If we are initializing sending flows here, why do you need add_sending_flow()?

adbh94 replied:
Some of our flows may not send data from the start of the simulation. I'm not sure how this will be implemented, so I included the add_sending_flow() method in case the environment was going to add those flows later in the simulation. This may not be needed if we program the flows to start sending data at a particular time, but it certainly won't hurt.

kexinrong commented on host.py:
```diff
@@ -0,0 +1,170 @@
+                new_receiving_flow = ReceivingFlow(flow_id)
+                self.flows[flow_id] = new_receiving_flow
+                new_receiving_flow.receive_packet(incoming_packet)
+
```
Here we are handling the case for receiving data packets right? What about acknowledgement packets? Does the receiving flow trigger the receive_ack event in the sending flow directly? If that's the case, at least the packets should have a field for sending flow id as well. Right now incoming_packet.get_flow_id() is only storing receiving flow id.

adbh94 replied:
I might not have understood your question fully, but here's my rationale on the monitor_incoming_packets() method. For definitions, a "sending_flow" sends data packets and receives acknowledgements, while a "receiving_flow" receives data packets and sends acknowledgements.

The key is that a sending_flow and its corresponding receiving_flow share the same flow_id. That way, a host can simply look at the flow_id parameter on an incoming packet and forward it to the appropriate flow regardless of the packet type. For example, if the host gets an acknowledgement packet, it will just check the flow_id of the packet and pass it off to the sending_flow with the matching parameter. There is never any telepathy going on between sending_flows and receiving_flows.

kexinrong replied:
Sorry my bad I didn't realize that the corresponding sending and receiving flow have the same id. It makes sense then.

adbh94 committed:
Bug fix.

Did not pass environment argument correctly to internal processes.

adbh94 commented:
Yes, good call. Fixed in the last commit.

aman-agarwal commented on host.py:
```diff
@@ -0,0 +1,170 @@
+    
+            # Otherwise create a new receiving flow on-the-fly. 
+            else:
+                new_receiving_flow = ReceivingFlow(flow_id)
```
RecevingFlow should take parameters flow_id, source host object and destination host id. (I am going modify flow.py accordingly)

adbh94 committed:
ID changes

Changed get_host_id() method to get_id() to unify ID accessor interface.
Also, remove_flow() now takes a flow_id rather than a flow as argument.

adbh94 committed:
Removed add_sending_flow() method.

We've agreed to program flows to start at the appropriate time. Thus,
this method is not needed since the environment is not responsible for
inserting flows at the correct time.

adbh94 committed:
Changed call to link's enqueue method.

The link's enqueue function accepts packets from hosts/routers. Its
transmit function, however, is internal to the link and does the actual
transmission of packets across the wire.

adbh94 commented:
@kexinrong I changed the host's calls to add packets to the link buffer, so you can rename the corresponding method to be enqueue() rather than transmit(). Then the process doing the actual transmitting can be called transmit().

adbh94 committed:
Matched ReceivingFlow API

adbh94 committed:
Added statistics reporting function

adbh94 committed:
Host reports send/receive rate.

Added report() method that conveys the average per-host send/receive
rate since the last report.

Measurements are given in dictionary format. I rely on an enum of
Measurements that the environment expects from all its devices.

adbh94 commented on host.py:
```diff
@@ -0,0 +1,197 @@
+        self.num_packets_received = 0.0
+        
+        return {self.env.Measurements.host_send_rate : host_send_rate,
+                self.env.Measurements.host_receive_rate : host_receive_rate}
```
I'm relying on an enum of all expected Measurements provided by the environment. We can then standardize our reports as dictionaries with keys identifying the metric from the enum and values being the measurements themselves.

adbh94 committed:
Reordered initialization.

kexinrong commented on host.py:
```diff
@@ -0,0 +1,197 @@
+        MS_TO_S = 1000
+
+        # Amount of time elapsed in seconds since the last report.
+        time_interval = (self.env.now - self.interval_start_time) * MS_TO_S
```
Can we just use self.env.interval instead?

adbh94 replied:
Done

adbh94 committed:
Use environment time interval

Rather than measuring time interval internally, host uses environment's
interval attribute that specifies how often it will call on report.

adbh94 committed:
Corrected ms to s conversion

adbh94 committed:
Fixed initialization and conversion factors

Forgot to use self keyword during initialization and used wrong
milliseconds to seconds conversion factor.

sdhpzhtk commented:
So the first events or the starting events of a host in the main environment are those two monitor functions?

adbh94 commented:
The host actually first creates two unscheduled events send_packet and receive_packet. After that, the two monitor functions comprise the host process. They are started when initializing the host and keep running until the simulation is terminated.

Note that the host object sets up these events in the environment by itself, so all the environment needs to do is create a host object.

adbh94 committed:
Changed initialization and added mutators.

Made dependent link and flow objects optional during initialization with
default values. Added mutators for environment to populate these
parameters after initialization.

adbh94 committed:
Reordered mutators for aesthetic purposes.

sdhpzhtk commented on host.py:
```diff
@@ -0,0 +1,202 @@
+        self.num_packets_sent = 0.0
+        self.num_packets_received = 0.0
+        
+        return {self.env.Measurements.host_send_rate : host_send_rate,
```
Please do not use an enum type.... just use a string as key, and I feel like it is fine.... We can discuss it tmr tho.

aman-agarwal replied:
@adbh94 @sdhpzhtk Decide this so that we can merge soon. I am okay either way. @kexinrong Do you have an opinion on this?

aman-agarwal commented:
In the report function, you should call the constant 0.001 MS_TO_S.

aman-agarwal commented:
Other than the report function, this branch looks ready to merge to me.

adbh94 committed:
Matched ReceivingFlow constructor API

adbh94 committed:
Match report() API in environment.

adbh94 committed:
Name change from receiving_flow to ReceivingFlow
